### DISCUSSION TITLE
Deterministic reset

### DISCUSSION THREAD

**Discussion by LeXXik** [jrouwe/JoltPhysics.js#225] [Q&A] [answered]
So, I'm trying to make a little bowling game. I am sending limited data over network, so I'd like to have the simulation deterministic.

In the clip: the ball starts to roll as kinematic (blue), then switches to dynamic (yellow) just before hitting the pins. The physics system has a truthy `mDeterministicSimulation`. On every reset:
- All bodies are destroyed (including static), and everything is created anew in the same order.
- The switch from kinematic to dynamic is happening at exactly the same world point (console logs xyz in the clip).
- The ball velocity is set to exactly the same one after the switch to dynamic.
- The pins are placed at the same hardcoded positions.

https://github.com/user-attachments/assets/77ca5fad-2e8f-4536-bd06-36994d32e85a

However, I can't make them fall deterministically. Am I missing something? I am hoping I don't need to destroy the Jolt instance and just adding/removing bodies from the physics world would suffice.

While at it - notice one pin keeps wobbling. Its angular damping is 0.1, which doesn't seem to help. Any other way to make it calm down sooner? Each pin is a compound collider made of 2 convex hulls.

PS
Merry Christmas and a Happy New Year!

**Comment by LeXXik** [comment]
Hmm, something with pins positions. When I spawn a dynamic pin, I set its position to some hardcoded vector. However, when the ball is about to hit it, the position is a bit different. I suppose the pin settles a bit after a spawn. That settling is different every time, though.

For example, I spawn a pin at a fixed position. When the ball is about to hit it, I read the pin's position as:

```
fixed: [15.629298210144043, -0.0009704494150355458, 0.7499359846115112]

reset 1: [15.62933349609375, -0.001925505232065916, 0.7498801350593567]
reset 2: [15.629328727722168, -0.0019221511902287602, 0.7498792409896851]
reset 3: [15.6293306350708, -0.0019232442136853933, 0.7498795390129089]
```

Pretty sure this is the cause of it, but not sure what is causing it.

**Comment by jrouwe** [comment, marked as answer]
Hello,

Please take a look at the BodyIDs of of your bodies. The values of those IDs determine the sort order and are essential for determinism. Unless you manage your own BodyIDs (through e.g. `BodyInterface::CreateBodyWithID`), I don't think it is possible to get the system to return the same BodyIDs again (because the sequence number will increment). This means that the simulation will not be the same if you delete and recreate all bodies using the same `PhysicsSystem`. Alternatively, you create a new `PhysicsSystem` for every run to get the same BodyIDs.

> While at it - notice one pin keeps wobbling. Its angular damping is 0.1, which doesn't seem to help. Any other way to make it calm down sooner? Each pin is a compound collider made of 2 convex hulls.

Hard to say. Ensure that the convex radius is big enough, that you haven't set `PhysicsSettings::mPenetrationSlop` to zero, that you let the system calculate the inertia for you.

> Merry Christmas and a Happy New Year!

Happy new year to you too!

**Reply by LeXXik** [reply]
Hmm, I am trying to use a custom body ID, but the method doesn't seem to return a body:

```
const body = bodyInterface.CreateBodyWithID(new Jolt.BodyID(0), bodyCreationSettings); // undefined
```
Looks like no return?

![image](https://github.com/user-attachments/assets/48e13f8e-2cea-45db-ba90-9a01817dd718)

**Reply by LeXXik** [reply]
Ah, maybe I need to update Jolt

**Reply by jrouwe** [reply]
My bad, it's a missing return type. Fixed it in: https://github.com/jrouwe/JoltPhysics.js/commit/4b9b7c22555483a6bb1647aa89c2ccedb91c5441

**Reply by LeXXik** [reply]
Meanwhile, I want to try resetting the physics system. What would be a correct way to destroy/create it? We don't seem to instantiate it, but get from the jolt interface. Should I `Jolt.destroy()` existing interface, create a new one and then get the physics system from it?

Edit: and I suppose I need to destroy all existing objects, like bodies/constraints myself before destroying the interface?

**Reply by jrouwe** [reply]
The safest way is to `Jolt.destroy` the `Jolt.JoltInterface` and, prior to that, all objects that were associated with it (shapes, bodies, constraints etc.). Technically you could probably keep some objects alive (e.g. shapes should work), but the easiest thing is just to remove and recreate everything.

**Reply by LeXXik** [reply]
Alright, I am destroying all objects with interface last. Then create all anew (frontend assigns a new component id after reset, can be seen in console log). Still the position of the first bowling pin is different between resets. The body id seems to be the same though (if I understand it correctly). I will try creating the bodies with id, once the patch is live.

![image](https://github.com/user-attachments/assets/2f912a24-9f06-4485-8cce-50a9ba622229)

**Reply by jrouwe** [reply]
Are you running the simulation with a constant delta time? (fluctuations in fps can make a difference in the simulation)

**Reply by LeXXik** [reply]
Yes, I call step() with a fixed timestep value.

**Reply by jrouwe** [reply]
I've released 0.31.0 with the fix for CreateBodyWithID.

It's hard for me to say something sensible about why the sim is different the 2nd time you run it. This is something that [PerformanceTest](https://github.com/jrouwe/JoltPhysics/tree/master/PerformanceTest) does too and it is used to validate determinism. It also destroys the physics system and recreates it for every test and it produces consistent hashes over many runs:

```
cd Build
./cmake_linux_emscripten.sh Distribution
cd WASM_Distribution
make
node ./PerformanceTest.js -q=LinearCast -s=Ragdoll -t=max -repeat=10
```

outputs:

```
Single precision WASM 32-bit with instructions: (16-bit ObjectLayer) (ObjectStream)
Running scene: Ragdoll
Motion Quality, Thread Count, Steps / Second, Hash
LinearCast, 1, 39.690629, 0x275057ded572c916
LinearCast, 1, 42.284898, 0x275057ded572c916
LinearCast, 1, 42.270048, 0x275057ded572c916
LinearCast, 1, 40.511347, 0x275057ded572c916
LinearCast, 1, 42.645867, 0x275057ded572c916
LinearCast, 1, 41.571807, 0x275057ded572c916
LinearCast, 1, 40.993676, 0x275057ded572c916
LinearCast, 1, 42.604562, 0x275057ded572c916
LinearCast, 1, 40.256976, 0x275057ded572c916
LinearCast, 1, 42.166087, 0x275057ded572c916
```

Is it the same every time you do a 1st run?

**Reply by LeXXik** [reply]
Thank you for the performance test! I've checked the CPP version, on whether I do the same steps. I will build the JS variant as you showed later for comparison. I probably still missed something.

I've also recreated settings, and broadphase filters, as it looks like I can't re-use the old ones.

```js
Jolt.destroy(joltInterface);
Jolt.destroy(bpFilter);
Jolt.destroy(objFilter);

const settings = new Jolt.JoltSettings();
settings.mObjectLayerPairFilter = new Jolt.ObjectLayerPairFilterTable(n);
settings.mBroadPhaseLayerInterface = ...;

const newInterface = new Jolt.JoltInterface(settings);
Jolt.destroy(settings);

const systemSettings = physicsSystem.GetPhysicsSettings();
const physicsSystem = newInterface.GetPhysicsSystem();

systemSettings.mBaumgarte = ...;

physicsSystem.SetPhysicsSettings(systemSettings);

const bpFilter = new Jolt.DefaultBroadPhaseLayerFilter(newInterface.GetObjectVsBroadPhaseLayerFilter(), 1);
const objFilter = new Jolt.DefaultObjectLayerFilter(newInterface.GetObjectLayerPairFilter(), 1);
```

I now do get X and Y the same values over multiple runs and browser refresh. But Z is still changing, so something is still amiss.

![image](https://github.com/user-attachments/assets/49e418a5-f129-415e-b22d-7dcfab3a8f96)

**Reply by LeXXik** [reply]
I see, the JS version is simply accessing the CPP one from Wasm. Can't use it to compare the code.

**Reply by LeXXik** [reply]
Tried using ID's, without resetting - getting the same result as in my previous post, where X and Y are the same, and Z is changing. Hmm, strange pattern. I will keep digging.

**Reply by LeXXik** [reply]
Yep, found it. The issue was in dt. Well, not the dt that I pass to Jolt, but an accumulator, that accumulates dynamic dt and is used to step the world later. That accumulator was not clamped, so the world could step once or twice, depending on accumulated value. Once clamped, though, I do get deterministic both with resetting the physics system and without (using custom ids). Thank you!

![image](https://github.com/user-attachments/assets/93614d01-2a46-4946-92f8-58b866c38c95)

> Ensure that the convex radius is big enough, that you haven't set PhysicsSettings::mPenetrationSlop to zero, that you let the system calculate the inertia for you.

It looks like increasing the penetration slop, helped! Thank you! Increasing convex radius didn't seem to change the wobble.

**Comment by LeXXik** [comment]
A related question, don't want to create a new thread. Does the step count matter for a deterministic sim? Say, in one run the dynamic objects spread at step 100, on another at step 150 (with all other params been the same).

**Reply by jrouwe** [reply]
When all objects are sleeping until frame 100/150 then I think the runs should be the same (in this case PhysicsSystem::Update is basically a no-op). In all other cases I think you run the risk of a desync (e.g. because the internal sleep timers are out of sync).
